sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py

```diff
@@ -764,3 +748,146 @@ def update_repository_properties(self, *args, **kwargs):
                 repository, value=properties._to_generated(), **kwargs  # pylint: disable=protected-access
             )
         )
+
+    @distributed_trace
+    def upload_manifest(
+        self, repository: str, manifest: "Union['OCIManifest', 'IO']", *, tag: "Optional[str]" = None, **kwargs: "Any"
+    ) -> None:
+        """Upload a manifest for an OCI artifact.
+
+        :param str repository: Name of the repository
+        :param manifest: The manifest to upload.
+        :type manifest: ~azure.containerregistry.models.OCIManifest or IO
+        :keyword tag: Tag of the manifest.
+        :paramtype tag: str or None
+        :returns: None
+        :rtype: None
+        :raises ValueError: If the parameter repository or manifest is None.
+        """
+        if (repository or manifest) is None:
+            raise ValueError("The parameter repository and manifest cannot be None.")
+
+        tag_or_digest = tag
+        if tag is None:
+            stream = manifest
+            if isinstance(manifest, OCIManifest):
+                stream = _serialize_manifest(manifest)
+            tag_or_digest = _compute_digest(stream)
+
+        self._client.container_registry.create_manifest(
+            name=repository, reference=tag_or_digest, payload=stream, content_type=OCI_MANIFEST_MEDIA_TYPE, **kwargs
+        )
+
+    @distributed_trace
+    def upload_blob(self, repository, stream, **kwargs):
+        # type: (str, IO, **Any) -> None
+        """Upload an artifact blob.
+
+        :param str repository: Name of the repository
+        :param stream: The manifest to upload.
+        :type stream: IO
+        :returns: None
+        :rtype: None
+        :raises ValueError: If the parameter repository or stream is None.
+        """
+        if (repository or stream) is None:
+            raise ValueError("The parameter repository and stream cannot be None.")
+
+        start_upload_respose = self._client.container_registry_blob.start_upload(
```

Comment:
Linking together 3 REST calls into a single method can have complications - is this `upload_blob` method idempotent? If the `complete_upload` call on line 803 fails due to flaky network and I call `upload_blob` a second time with the same inputs - will it succeed?

Reply:
The service supports both chunked and resumable/range APIs here. We have decided to use chunked w/o range so we basically upload everything as a single blob. Every API call here can be retried. The idea is to scope it currently to only support wasteful upload (every time we start afresh if something goes wrong) however as we get more customers, we can use the range feature and become smart about it.